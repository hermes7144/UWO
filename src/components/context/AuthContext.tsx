import { createContext, useContext } from 'react';
import { User } from '@firebase/auth';

export const AuthContext = createContext<User | null>(null);

export function useAuthContext() {
  return useContext(AuthContext);
}
