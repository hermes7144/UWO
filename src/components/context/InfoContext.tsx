import { createContext, useContext } from 'react';

interface IInfoContext {
  getGoods(city: string): Promise<any[]>;
}

export const InfoContext = createContext<IInfoContext | null>(null);

export function useInfoContext() {
  return useContext(InfoContext);
}
