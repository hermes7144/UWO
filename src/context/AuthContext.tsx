import { createContext, useContext } from 'react';
import { User } from '@firebase/auth';

type UserType = {
  user: User & {
    isAdmin?: boolean;
  };
  uid?: string;
  login: () => {};
  logout: () => void;
};

export const AuthContext = createContext<UserType | null>(null);

export function useAuthContext() {
  return useContext(AuthContext);
}
