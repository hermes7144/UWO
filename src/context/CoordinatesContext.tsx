import { createContext, useContext } from 'react';

interface ICoordinatesContext {
  coordinates: [number, number];
  setCoordinates: React.Dispatch<React.SetStateAction<[number, number]>>;
}

export const CoordinatesContext = createContext<ICoordinatesContext | null>(null);

export function useCoordinatesContext() {
  return useContext(CoordinatesContext);
}
