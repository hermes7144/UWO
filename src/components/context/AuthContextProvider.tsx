import React from 'react';
import { useEffect, useState } from 'react';
import { login, logout, onUserStateChange } from '../../api/firebase';
import { AuthContext } from './AuthContext';
import { User } from '@firebase/auth';

type AuthContextProviderProps = {
  children: React.ReactNode;
};

export function AuthContextProvider({ children }: AuthContextProviderProps) {
  const [user, setUser] = useState<User | null>(null);

  useEffect(() => {
    onUserStateChange((user: User) => {
      setUser(user);
    });
  }, []);

  return <AuthContext.Provider value={{ user, uid: user && user.uid, login, logout }}>{children}</AuthContext.Provider>;
}
