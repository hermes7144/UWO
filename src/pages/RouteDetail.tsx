import React from 'react';
import Button from '../components/ui/Button';
import { useNavigate, useParams } from 'react-router-dom';
import useRoute from '../Hooks/useRoute';
import { useAuthContext } from '../context/AuthContext';
import Map from '../components/Map';
import Citys from '../components/Citys';

export default function RouteDetail() {
  const navigate = useNavigate();
  const { id } = useParams();
  const { uid } = useAuthContext();
  const { removeItem } = useRoute();

  const {
    routeQuery: { isLoading, error, data: route },
  } = useRoute();

  const handleUpdate = () => {
    navigate(`/routes/update/${id}`, { state: { route } });
  };

  const handleDelete = () => {
    if (!window.confirm('삭제하시겠습니까?')) return false;

    removeItem.mutate(
      { id },
      {
        onSuccess: () => {
          navigate('/routes');
        },
      }
    );
  };
  return (
    <>
      {isLoading && <p>Loading...</p>}
      {error && <p>Error</p>}
      {route && (
        <div className='flex flex-col sm:flex-row'>
          <div className='basis-4/6'>
            <Map citys={route.citys} isEditable={false} onMarker={() => {}} />
          </div>
          <div className='basis-2/6 flex flex-col p-2'>
            {route.user_id === uid && (
              <div className='flex justify-end font-semibold'>
                <Button text={'수정'} onClick={handleUpdate} />
                <Button text={'삭제'} onClick={handleDelete} />
              </div>
            )}
            <h1 className='text-2xl'>{route.title}</h1>
            <span className='whitespace-pre'>{route.description}</span>
            <Citys citys={route.citys} major_goods={route.major_goods} major_chk={route.major_chk} isEditable={false} onDelete={() => {}} />
          </div>
        </div>
      )}
    </>
  );
}
