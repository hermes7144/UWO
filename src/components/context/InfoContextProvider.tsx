import { getGoods } from '../../api/firebase';
import { InfoContext } from './InfoContext';
import { useState } from 'react';

type ContextProviderProps = {
  children: React.ReactNode;
};
export function InfoContextProvider({ children }: ContextProviderProps) {
  const [citys, setCitys] = useState<number[]>([]);

  type RouteType = {
    id: string;
    title: string;
    remark: string;
  };
  const [route, setRoute] = useState<RouteType>({ id: '', title: '', remark: '' });

  return <InfoContext.Provider value={{ citys, setCitys, route, setRoute, getGoods }}>{children}</InfoContext.Provider>;
}
