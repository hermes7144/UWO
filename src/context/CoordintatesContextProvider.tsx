import { CoordinatesContext } from './CoordinatesContext';
import { useState } from 'react';

type ContextProviderProps = {
  children: React.ReactNode;
};
export function CoordinatesContextProvider({ children }: ContextProviderProps) {
  const [coordinates, setCoordinates] = useState<[number, number]>();

  return <CoordinatesContext.Provider value={{ coordinates, setCoordinates }}>{children}</CoordinatesContext.Provider>;
}
