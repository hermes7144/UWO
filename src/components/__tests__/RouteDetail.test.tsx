import { render, screen } from '@testing-library/react';
import { Route } from 'react-router-dom';
import { useAuthContext } from '../../context/AuthContext';
import RouteDetail from '../../pages/RouteDetail';
import { fakeMarkers } from '../../tests/fakeGoods';
import { fakeRoute } from '../../tests/fakeRoutes';
import { withAllContexts, withRouter } from '../../tests/utils';

jest.mock('../../context/AuthContext');

describe('Router Detail', () => {
  const mockUseAuthContext = useAuthContext as jest.MockedFunction<typeof useAuthContext>;
  const fakeUseRoute = jest.fn();

  beforeEach(() => {
    mockUseAuthContext.mockReturnValue({
      user: null,
      login: jest.fn(),
      logout: jest.fn(),
      uid: 'uid',
    });

    fakeUseRoute.mockImplementation(() => {
      return {
        routeQuery: { isLoading: false, data: fakeRoute },
        markersQuery: { isLoading: false, data: fakeMarkers },
      };
    });
  });

  afterEach(() => fakeUseRoute.mockReset());

  it('renders correctly', async () => {
    const { asFragment } = renderRouterDetail();

    await screen.findAllByText('title');
    expect(asFragment()).toMatchSnapshot();
  });

  function renderRouterDetail() {
    return render(withAllContexts(withRouter(<Route path='/' element={<RouteDetail />} />), fakeUseRoute));
  }
});
