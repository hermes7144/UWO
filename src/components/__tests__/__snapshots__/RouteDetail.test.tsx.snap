// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router Detail renders correctly 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col p-2"
  >
    <div
      class="flex justify-end font-semibold"
    >
      <button
        class="bg-brand text-white
 py-2 px-4 rounded-sm hover:brightness-110"
      >
        수정
      </button>
      <button
        class="bg-brand text-white
 py-2 px-4 rounded-sm hover:brightness-110"
      >
        삭제
      </button>
    </div>
    <h1
      class="text-2xl"
    >
      title
    </h1>
    <span
      class="whitespace-pre"
    >
      description
    </span>
    <div />
  </div>
</DocumentFragment>
`;
